src/main/rx/srxl2.c

```diff
@@ -0,0 +1,610 @@
+      readBufferReadIdx = readBufferIdx;
+      globalResult = RX_FRAME_DROPPED;
+      return;
+    }
+
+    //    cliPrintf("> %d %d %x\r\n", readBufferIdx, readBufferReadIdx, srxl2Frame.header.id);
+
+    const uint8_t bytesToRead = srxl2Frame.header.length;
+    if (bytesToRead > sizeof(srxl2Frame)) {
+      readBufferReadIdx = readBufferIdx;
+      globalResult = RX_FRAME_DROPPED;
+      return;
+    }
+
+    while (bytesCopied < bytesToRead) {
+        srxl2Frame.localBuffer[bytesCopied++] = readBuffer[readBufferReadIdx++];
+
+        if (readBufferReadIdx == sizeof(readBuffer)) {
+            readBufferReadIdx = 0;
+        }
+    }
+
+    //    cliPrintf("- %d %d %x\r\n", readBufferIdx, readBufferReadIdx, srxl2Frame.header.id);
+
+    if (bytesCopied != srxl2Frame.header.length) {
+        readBufferReadIdx = readBufferIdx;
+        globalResult = RX_FRAME_DROPPED;
+        return;
+    }
+
+    const uint16_t calculatedCrc = crc16_ccitt_update(0, srxl2Frame.localBuffer, srxl2Frame.header.length - 2);
+    //    cliPrintf("< %d %d %x %x\r\n", readBufferIdx, readBufferReadIdx, srxl2Frame.header.id, calculatedCrc);
+
+    const uint16_t receivedCrc =
+        (srxl2Frame.localBuffer[srxl2Frame.header.length - 2] << 8) |
+        (srxl2Frame.localBuffer[srxl2Frame.header.length - 1]);
+
+    if (calculatedCrc != receivedCrc) {
+        readBufferReadIdx = readBufferIdx;
+        globalResult = RX_FRAME_DROPPED;
+        //        DEBUG("crc mismatch %x vs %x\r\n", calculatedCrc, receivedCrc);
+        return;
+    }
+
+    //Packet is valid only after ID and CRC check out
+    lastValidPacketTimestamp = micros();
+
+    if (srxl2ProcessPacket(&srxl2Frame.header, rxRuntimeConfig)) {
+        return;
+    }
+
+    // @todo reset
+    //    DEBUG("could not parse packet: %x\r\n", srxl2Frame.header.packetType);
+    globalResult = RX_FRAME_DROPPED;
+}
+
+
+static void srxl2DataReceive(uint16_t character, void *data)
+{
+    UNUSED(data);
+
+    lastReceiveTimestamp = microsISR();
+
+    // circular buffer depleted and waiting for sync character
+    if (readBufferReadIdx == readBufferIdx && character != SRXL2_ID) {
+        return;
+    }
+
+    readBuffer[readBufferIdx] = character;
+    readBufferIdx++;
+    if (readBufferIdx == sizeof(readBuffer)) {
+      readBufferIdx = 0;
+    }
+}
+
+static void srxl2Idle()
+{
+    lastIdleTimestamp = microsISR();
+}
+
+uint32_t autobaud_timeout = 0;
+static uint8_t srxl2FrameStatus(rxRuntimeConfig_t *rxRuntimeConfig)
+{
+    UNUSED(rxRuntimeConfig);
+
+    globalResult = RX_FRAME_PENDING;
+
+    const uint32_t bytesAvailable =
+        readBufferReadIdx > readBufferIdx
+            ? readBufferIdx + (uint8_t) sizeof(readBuffer) - readBufferReadIdx
+            : readBufferIdx - readBufferReadIdx;
+
+    if (bytesAvailable >= sizeof(Srxl2Header)) {
+        union {
+            uint8_t localBuffer[sizeof(Srxl2Header)];
+            Srxl2Header header;
+        } srxl2Frame;
```

Comment:
This union is defined locally in a number of functions - having one unit wide definition will make this cleaner.

Reply:
Double buffer changes have defined a modified version of this union globally, and used it to declare the buffers. Please verify if this is acceptable.